src/vsr/replica_format.zig

```diff
@@ -26,195 +22,287 @@ pub fn format(
 
     var replica_format = ReplicaFormat{};
 
-    try replica_format.format_wal(allocator, options.cluster, storage);
-    assert(!replica_format.formatting);
-
-    try replica_format.format_replies(allocator, storage);
-    assert(!replica_format.formatting);
+    var arena = std.heap.ArenaAllocator.init(allocator);
+    defer arena.deinit();
 
-    try replica_format.format_grid_padding(allocator, storage);
-    assert(!replica_format.formatting);
+    try replica_format.queue_format_wal(arena.allocator(), options.cluster, storage);
+    try replica_format.queue_format_replies(arena.allocator(), storage);
+    try replica_format.queue_format_grid_padding(arena.allocator(), storage);
 
-    superblock.format(
-        ReplicaFormat.format_superblock_callback,
-        &replica_format.superblock_context,
-        options,
-    );
+    replica_format.format_and_tick(storage, &superblock, options);
 
-    replica_format.formatting = true;
-    while (replica_format.formatting) storage.run();
+    assert(!replica_format.formatting and !replica_format.formatting_superblock);
+    assert(replica_format.writes_pending == 0);
 }
 
+/// When formatting, we write:
+/// * constants.journal_slot_count many prepares,
+/// * 1 write that contains all of the headers,
+/// * constants.clients_max many replies,
+/// * 1 block for grid padding, maybe.
+pub const writes_max = constants.journal_slot_count + 1 + constants.clients_max + 1;
+
 fn ReplicaFormatType(comptime Storage: type) type {
     const SuperBlock = vsr.SuperBlockType(Storage);
     return struct {
         const ReplicaFormat = @This();
+        var formating_instance: ?*ReplicaFormat = null;
```

Comment:
I do this sort of global mutable variable thing in `testing/fixtures.zig`, and, as that's testing code, that's probably.

But this is production code, so I think we shouldn't assume that there's only single global instance of formatting. Let's do the proper thing and define our own 

```zig
const Write = struct {
    write: Storage.Write, 
    replica_format: *ReplicaFormat,
}
```